### DISCUSSION TITLE
Is it possible to group packages with captured patterns in package name?

### DISCUSSION THREAD

**Discussion by lyonlai** [renovatebot/renovate#27721] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

renovate/renovate:37.219.4-slim

### If you're self-hosting Renovate, select which platform you are using.

GitHub Enterprise Server

### What is your question?

Not sure if it make sense for other package managers. In Gradle, dependency name structure is `{{group}}:{{name}}:{{version}}`. 

Not sure if it is possible now to organise dependencies by their group name. So that we have modules publish from the same repository (usually that's when the {{group}} is the same) grouped into a single PR.

### Logs (if relevant)

<details><summary>Logs</summary>

```

Copy/paste the relevant log(s) here, between the starting and ending backticks

```

</details>

**Comment by rarkins** [comment]
npm similar has an optional "scope", e.g. `@renovate/abc` where `@renovate` is the scope part.

Scope may be preferable to use over groupId considering we already use the term "group" in Renovate a lot.

I'm not sure where this scope identification should be done though. It seems like the Maven and npm managers could identify their scopes pretty easy during extraction, but it might not be so simple for others. And for example if npm has github dependencies we probably want the github org name to be the scope, and suddenly npm then needs to be github-aware. So I'm wondering if the scope is more associated with datasource than manager, internally.

**Reply by lyonlai** [reply]
extraction might be one place to do so. But as you said sometimes datasource needs to be taken into consideration and there might be other things to consider down the road which will complicate what manager does.

The other place I'm thinking is maybe we could make `groupName` and `groupSlug` be able to pick up regex groups captured with  `matchPackagePatterns` or `matchDepPatterns`.  what do you think?

**Reply by rarkins** [reply]
It's a possibility, but it sounds like a level of complexity I'd like to avoid if there's a simpler way for this common use case

**Reply by lyonlai** [reply]
how about two attributes and use them together?

* an attribute `captureGroup` or `matchGroups` that apply a regex on packageName. Like [`matchStrings`](https://docs.renovatebot.com/configuration-options/#matchstrings) with capture groups defined
* `groupNameTemplate` picks those groups up like [`depNameTemplate`](https://docs.renovatebot.com/configuration-options/#autoreplacestringtemplate) in `customManagers`

**Reply by rarkins** [reply]
I would prefer a solution where datasources optionally return a "packageScope" value which can then be used in templates

**Reply by lyonlai** [reply]
Found [this line](https://github.com/renovatebot/renovate/blob/ea0f044047e1149a32e06ee2b02b337aacabc630/lib/workers/repository/updates/branch-name.ts#L51) while looking for a place for the `groupNameTemplate` logic. However, it seems like we are already compiling `groupName` with the `update` object. Does that mean the only missing bit in this feature is the `packageScope` from datasource?

**Reply by rarkins** [reply]
Could be!

**Reply by lyonlai** [reply]
Alright, I have played around locally with the [`replace` handlebar helper](https://github.com/renovatebot/renovate/blob/289944ed0aa2deb1d3c2b2715f392478c91daca0/lib/util/template/index.ts#L19-L21) in renovate and it managed to get the group name out as I wanted.

```json5
{
  packageRules: [
    {
      matchManagers: [
        "gradle"
      ],
      groupName: "{{replace ':.*$' '' depName}}"
    }
  ]
}
```

and the trace level log 

```
Dependency org.jetbrains.kotlinx:kotlinx-coroutines-rx2 is part of group org.jetbrains.kotlin
Dependency org.jetbrains.kotlinx:kotlinx-coroutines-core is part of group org.jetbrains.kotlinx

```

However, this shouldn't be the way forward long term. 

`packageScope` seems to be a proper way forward, which I have also managed to make it happen from `maven` datasource to the `groupName` template. And the changes are quite small. Would you like to see a PR for it?

**Reply by lyonlai** [reply, marked as answer]
[PR](https://github.com/renovatebot/renovate/pull/27925) is up. PTAL
